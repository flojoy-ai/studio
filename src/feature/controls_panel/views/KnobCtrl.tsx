import {
  CtlManifestType,
  CtrlManifestParam,
} from "@src/hooks/useFlowChartState";
import { Draft } from "immer";
import { memo, useCallback, useEffect, useState } from "react";
import Silver from "@src/utils/SilverKnob";
import ReactGridLayout from "react-grid-layout";
import { ControlOptions } from "../types/ControlOptions";
import { ParamValueType } from "@feature/common/types/ParamValueType";

export interface KnobCtrlProps {
  makeLayoutStatic: () => void;
  isEditMode: boolean;
  ctrlObj: CtlManifestType;
  selectedOption: ControlOptions | undefined;
  setGridLayout: (
    update:
      | ReactGridLayout.Layout[]
      | ((draft: Draft<ReactGridLayout.Layout>[]) => void)
  ) => void;
  updateCtrlValue: (
    value: string,
    ctrl: CtlManifestType,
    ValTyp: ParamValueType
  ) => void;
  currentInputValue: number;
}

const KnobCtrl = ({
  makeLayoutStatic,
  isEditMode,
  setGridLayout,
  ctrlObj,
  updateCtrlValue,
  currentInputValue,
  selectedOption,
}: KnobCtrlProps) => {
  const [knobValue, setKnobValue] = useState(0);
  const updateCtrlValueFromKnob = useCallback(
    (value: number) => {
      if ((ctrlObj?.param as CtrlManifestParam)?.nodeId) {
        updateCtrlValue(value.toString(), ctrlObj, "float");
      }
      setKnobValue(value);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [(ctrlObj?.param as CtrlManifestParam)?.nodeId]
  );

  useEffect(() => {
    return () => {
      setKnobValue(0);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedOption]);

  return (
    <div
      onMouseEnter={makeLayoutStatic}
      onMouseLeave={() => {
        if (isEditMode) {
          setGridLayout((prev) => {
            prev[prev.findIndex((layout) => layout.i === ctrlObj.id)].static =
              false;
          });
        }
      }}
      style={{
        width: "100%",
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
      }}
      data-testid={"KnobCtrlDiv"}
    >
      <Silver
        style={{
          width: "fit-content",
          boxShadow: "0",
          zIndex: 1,
        }}
        // diameter={70}
        knobStyle={{ boxShadow: "0", background: "none" }}
        min={0}
        max={100}
        step={1}
        value={knobValue || currentInputValue}
        diameter={180}
        onValueChange={updateCtrlValueFromKnob}
        ariaLabelledBy={"my-label"}
      ></Silver>
    </div>
  );
};

export default memo(KnobCtrl);
