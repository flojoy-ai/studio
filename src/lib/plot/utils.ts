export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
