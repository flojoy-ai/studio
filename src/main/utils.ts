import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
