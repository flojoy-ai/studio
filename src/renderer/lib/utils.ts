import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
