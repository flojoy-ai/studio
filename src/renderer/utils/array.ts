export function filterMap<T, U>(
  arr: Array<T>,
  mapPred: (x: T) => U | undefined | null,
) {
  return arr.reduce((acc: Array<U>, x: T) => {
    const val = mapPred(x);
    if (val !== undefined && val !== null) {
      acc.push(val);
    }

    return acc;
  }, []);
}
