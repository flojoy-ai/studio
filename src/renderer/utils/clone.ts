import { DeepMutable } from "@/renderer/types/util";

export const deepMutableClone = <T>(obj: T): DeepMutable<T> => {
  return JSON.parse(JSON.stringify(obj)) as DeepMutable<T>;
};
