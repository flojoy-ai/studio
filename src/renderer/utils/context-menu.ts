export const calculateContextMenuOffset = (
  clickX: number,
  clickY: number,
  flow: HTMLDivElement,
) => {
  // Calculate position of the context menu. We want to make sure it
  // doesn't get positioned off-screen.
  const pane = flow.getBoundingClientRect();

  const contextMenuHeight = 200;
  const paneToBlock = clickY - contextMenuHeight;

  let top: number | undefined = undefined;
  let bottom: number | undefined = undefined;

  if (paneToBlock < contextMenuHeight / 2) {
    top = paneToBlock;
  } else if (paneToBlock < contextMenuHeight) {
    if (pane.height - paneToBlock < contextMenuHeight) {
      top = contextMenuHeight - paneToBlock;
    } else {
      top = paneToBlock;
    }
  } else if (pane.height - paneToBlock < contextMenuHeight) {
    top = undefined;
    bottom = pane.height - paneToBlock;
  } else {
    top = paneToBlock;
  }

  return {
    top,
    left: clickX < pane.width - 200 ? clickX : undefined,
    right: clickX >= pane.width - 200 ? pane.width - clickX : undefined,
    bottom,
  };
};
