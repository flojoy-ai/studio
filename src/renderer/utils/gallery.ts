import { GalleryData } from "@/renderer/types/gallery";

export const data: GalleryData = {
  Fundamentals: [
    {
      title: "Intro to Loops",
      description: "Generate a random number once",
      imagePath: "assets/appGallery/introToLoops.png",
      appPath: "loop",
      relevantNodes: [
        {
          name: "LOOP",
          docs: "https://docs.flojoy.ai/blocks/control-flow/#loops",
        },
      ],
      cloudDemoEnabled: true,
    },
    {
      title: "Butterworth Filter",
      description: "Apply a butterworth filter to an input signal",
      imagePath: "assets/appGallery/introToSignals.png",
      appPath: "butterworth",
      relevantNodes: [],
      cloudDemoEnabled: true,
    },
    {
      title: "Intro to Images",
      description: "Apply a butterworth filter on a sample image",
      imagePath: "assets/appGallery/introToImages.png",
      appPath: "images",
      relevantNodes: [
        {
          name: "IMAGE",
          docs: "https://docs.flojoy.ai/blocks/data/visualization/plotly/image",
        },
      ],
      cloudDemoEnabled: true,
    },
  ],
  Oscilloscopes: [
    {
      title: "Tektronix",
      description: "Decode I2C messages with a MSO2X oscilloscope",
      imagePath: "assets/appGallery/MS024DecodeI2C.png",
      appPath: "MSO24DecodeI2C",
      relevantNodes: [
        {
          name: "DECODE_I2C_MSO2X",
          docs: "https://docs.flojoy.ai/blocks/hardware/oscilloscopes/tektronix/mso2x/decode-i2c-mso2x/",
        },
      ],
      cloudDemoEnabled: false,
    },
    {
      title: "Tektronix",
      description: "Extract the trace from a MDO3XXX oscilloscope",
      imagePath: "assets/appGallery/mdo3ExtractTrace.png",
      appPath: "mdo3ExtractTrace",
      relevantNodes: [
        {
          name: "EXTRACT_TRACE_MDO3XXX",
          docs: "https://docs.flojoy.ai/blocks/hardware/oscilloscopes/tektronix/mdo3xxx/extract-trace-mdo3xxx/",
        },
      ],
      cloudDemoEnabled: false,
    },
    {
      title: "Rigol",
      description: "Extract a trace from a Rigol oscilloscope",
      imagePath: "assets/appGallery/rigolExtractTrace.png",
      appPath: "rigolExtractTrace",
      relevantNodes: [
        {
          name: "EXTRACT_TRACE_DS1074Z",
          docs: "https://docs.flojoy.ai/instruments-database/oscilloscopes/rigol/rigol-ds1074z/",
        },
      ],
      cloudDemoEnabled: false,
    },
    {
      title: "LeCroy",
      description: "Extract a trace from a Teledyne LeCryo T3DSO oscilloscope",
      imagePath: "assets/appGallery/LeCroyT3DSO.png",
      appPath: "leCroyExtractTrace",
      relevantNodes: [
        {
          name: "CONNECT_T3DSO1XXX",
          docs: "https://docs.flojoy.ai/blocks/hardware/oscilloscopes/teledyne-lecroy/t3dso1xxx/connect-t3dso1xxx/",
        },
      ],
      cloudDemoEnabled: false,
    },
    {
      title: "PICO",
      description: "Extract the trace from a P2000 PicoScope",
      imagePath: "assets/appGallery/picoExtractTrace.png",
      appPath: "picoExtractTrace",
      relevantNodes: [
        {
          name: "EXTRACT_TRACE_2000",
          docs: "https://docs.flojoy.ai/blocks/hardware/oscilloscopes/pico/pico2000/extract-trace-2000/",
        },
      ],
      cloudDemoEnabled: false,
    },
  ],
  Function_Generators: [
    {
      title: "Bode Plot",
      description:
        "Create a Bode plot with a function generator and an oscilloscope",
      imagePath:
        "https://res.cloudinary.com/dhopxs1y3/image/upload/v1692118735/Instruments/Function%20Generators/AFG3000/AFG3000.png",
      appPath: "bodePlot",
      relevantNodes: [
        {
          name: "INPUT_PARAM_AFG31000",
          docs: "https://docs.flojoy.ai/blocks/hardware/function-generators/tektronix/afg31000/input-param-afg31000/",
        },
      ],
      cloudDemoEnabled: false,
    },
  ],
  DAQ: [
    {
      title: "LabJack",
      description: "Record and log temperatures",
      imagePath: "assets/appGallery/labjack.png",
      appPath: "labjack",
      relevantNodes: [
        {
          name: "READ_ANALOG_LABJACKU3",
          docs: "https://docs.flojoy.ai/blocks/hardware/daq-boards/labjack/u3/read-a0-pins/",
        },
      ],
      cloudDemoEnabled: false,
    },
    {
      title: "Arduino",
      description: "Read from any analog sensor",
      imagePath: "assets/appGallery/arduino.png",
      appPath: "arduino",
      relevantNodes: [],
      cloudDemoEnabled: false,
    },
    {
      title: "Read Analog value from NI CompactDAQ",
      description: "Record current value from a NI CompactDAQ",
      imagePath: "assets/appGallery/NIcDAQ.png",
      appPath: "cDAQReadAnalog",
      relevantNodes: [
        {
          name: "CONNECT_COMPACTDAQ",
          docs: "https://docs.flojoy.ai/blocks/hardware/daq-boards/national-instruments/compact-daq/create-task/",
        },
        {
          name: "ANALOG_INPUT",
          docs: "https://docs.flojoy.ai/blocks/hardware/daq-boards/national-instruments/compact-daq/attach-analog-input-current/",
        },
      ],
      cloudDemoEnabled: false,
    },
  ],
  Multimeters_Sourcemeters: [
    {
      title: "IV Sweep",
      description: "Perform an IV sweep",
      imagePath: "assets/appGallery/2450.png",
      appPath: "IVSweep",
      relevantNodes: [
        {
          name: "IV_SWEEP_2450",
          docs: "https://docs.flojoy.ai/blocks/hardware/sourcemeters/keithley/2450/iv-sweep-2450/",
        },
      ],
      cloudDemoEnabled: false,
    },
    {
      title: "Measure Voltage",
      description: "Read DC voltage using a digital multimeter",
      imagePath: "assets/appGallery/dmm.png",
      appPath: "dmmRead",
      relevantNodes: [
        {
          name: "READ_MEASUREMENT_DMM7510",
          docs: "https://docs.flojoy.ai/blocks/hardware/multimeters/keithley/dmm7510/read-measurement-dmm7510/",
        },
      ],
      cloudDemoEnabled: false,
    },
  ],
  Protocols: [
    {
      title: "Read CAN Bus",
      description:
        "Connect to a PEAK-USB device to capture CAN Bus messages and log them into a .mf4 file for analysis and storage",
      imagePath: "assets/appGallery/peak-usb.png",
      appPath: "canReadAndLog",
      relevantNodes: [
        {
          name: "CONNECT_PCAN",
          docs: "https://docs.flojoy.ai/blocks/hardware/protocols/can/bus/peak-connect/",
        },
        {
          name: "RECEIVE_MESSAGE",
          docs: "https://docs.flojoy.ai/blocks/hardware/protocols/can/bus/receive-can-message/",
        },
      ],
      cloudDemoEnabled: false,
    },
    {
      title: "Send CAN Bus Messages",
      description:
        "Connect to a CANable USB-CAN device and transmit frames onto a CAN Bus network.",
      imagePath: "assets/appGallery/canable-usb.png",
      appPath: "canSend",
      relevantNodes: [
        {
          name: "CONNECT_SLCAN",
          docs: "https://https://docs.flojoy.ai/blocks/hardware/protocols/can/bus/send-can-message/",
        },
        {
          name: "SEND_MESSAGE",
          docs: "https://https://docs.flojoy.ai/blocks/hardware/protocols/can/bus/send-can-message/",
        },
      ],
      cloudDemoEnabled: false,
    },
  ],
  Signal_Analyzers: [],
  IO: [
    {
      title: "USB Camera",
      description: "Capture real-time images",
      imagePath: "assets/appGallery/usbCamera.png",
      appPath: "webcam",
      relevantNodes: [
        {
          name: "OPEN_WEBCAM",
          docs: "https://docs.flojoy.ai/blocks/hardware/imaging/open-webcam/#_top",
        },
      ],
      cloudDemoEnabled: true,
    },
    {
      title: "Stepper Motor",
      description: "Precisely position anything",
      imagePath: "assets/appGallery/stepperMotor.png",
      appPath: "stepper",
      relevantNodes: [
        {
          name: "CONTROL_TIC_DRIVER",
          docs: "https://docs.flojoy.ai/blocks/hardware/motors/stepper/polulu/tic/",
        },
      ],
      cloudDemoEnabled: false,
    },
  ],
  DSP: [
    {
      title: "PID Controller",
      description: "Solve this non-linear dynamic system",
      imagePath: "assets/appGallery/PID.png",
      appPath: "pid",
      relevantNodes: [],
      cloudDemoEnabled: true,
    },
    {
      title: "FIR Filter",
      description: "Apply an FIR filter to an input signal",
      imagePath: "assets/appGallery/FIR.png",
      appPath: "fir",
      relevantNodes: [
        {
          name: "FIR",
          docs: "https://docs.flojoy.ai/blocks/dsp/fir/",
        },
      ],
      cloudDemoEnabled: true,
    },
    {
      title: "FFT",
      description: "Apply a real-time FFT to an input signal",
      imagePath: "assets/appGallery/FFT.png",
      appPath: "fft",
      relevantNodes: [
        {
          name: "FFT",
          docs: "https://docs.flojoy.ai/blocks/dsp/fft/",
        },
      ],
      cloudDemoEnabled: true,
    },
    {
      title: "IFFT",
      description: "Transform a signal with the IFFT algorithm",
      imagePath: "assets/appGallery/IFFT.png",
      appPath: "ifft",
      relevantNodes: [
        {
          name: "IFFT",
          docs: "https://docs.flojoy.ai/blocks/dsp/ifft/",
        },
      ],
      cloudDemoEnabled: true,
    },
  ],
  AI: [
    {
      title: "Image Captioning",
      description: "Caption any image with this PyTorch ML model",
      imagePath: "assets/appGallery/imageCaptioning.png",
      appPath: "imageCaptioning",
      relevantNodes: [
        {
          name: "NLP_CONNECT_VIT_GPT2",
          docs: "https://docs.flojoy.ai/blocks/ai-ml/image-captioning/nlp-connect-vit-gpt2",
        },
      ],
      cloudDemoEnabled: true,
    },
    {
      title: "Image Classification",
      description: "Classify any image using Hugging Face Transformers",
      imagePath: "assets/appGallery/imageClassification.png",
      appPath: "imageClassification",
      relevantNodes: [
        {
          name: "HUGGING_FACE_PIPELINE",
          docs: "https://docs.flojoy.ai/blocks/ai-ml/image-classification/hugging-face-pipeline/",
        },
      ],
      cloudDemoEnabled: true,
    },
    {
      title: "Time Series Forecasting",
      description: "Predict future events with the Prophet node",
      imagePath: "assets/appGallery/timeSeries.png",
      appPath: "prophet",
      relevantNodes: [
        {
          name: "PROPHET_PREDICT",
          docs: "https://docs.flojoy.ai/blocks/ai-ml/predict-time-series/prophet-predict/",
        },
      ],
      cloudDemoEnabled: true,
    },
  ],
};

export const getGalleryData = () => {
  return GalleryData.parse(data);
};
