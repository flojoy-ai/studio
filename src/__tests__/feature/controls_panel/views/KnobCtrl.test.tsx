import { renderWithTheme } from "@src/__tests__/__utils__/utils";
import KnobCtrl, {
  KnobCtrlProps,
} from "@src/feature/controls_panel/views/KnobCtrl";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const testKnobCtr: KnobCtrlProps = {
  makeLayoutStatic: jest.fn(),
  isEditMode: false,
  ctrlObj: {
    id: "plot-control",
    name: "Plot Control",
    type: "plot",
    minHeight: 10,
    minWidth: 10,
    layout: { i: "asd", x: 34, y: 34, w: 23, h: 3 },
    val: 30,
  },
  selectedOption: undefined,
  setGridLayout: jest.fn(),
  updateCtrlValue: jest.fn(),
  currentInputValue: 10,
};

jest.mock("@src/utils/SilverKnob", () => {
  return {
    __esModule: true,
    default: jest.fn().mockReturnValue(<></>),
  };
});

describe("KnobCtrl Component test", () => {
  it("KnobCtrl should render with Props", () => {
    const { container } = renderWithTheme(<KnobCtrl {...testKnobCtr} />);
    expect(container).toMatchSnapshot();
  });

  it("KnobCtrl should have hover effect", async () => {
    const user = userEvent.setup();
    renderWithTheme(<KnobCtrl {...testKnobCtr} />);

    await user.hover(screen.getByTestId("KnobCtrlDiv"));
    expect(testKnobCtr.makeLayoutStatic).toBeCalled();
  });
});
