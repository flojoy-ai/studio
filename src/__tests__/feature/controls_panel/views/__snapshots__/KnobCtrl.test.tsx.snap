// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KnobCtrl Component test KnobCtrl should render with Props 1`] = `
<div>
  <div
    data-testid="KnobCtrlDiv"
    style="width: 100%; display: flex; justify-content: center; align-items: center;"
  />
  ,
</div>
`;
